File: python/VCFOneParentProgenyImputed.py
from __future__ import annotations

# coding: utf-8
# VCFOneParentProgenyImputed.py
# 片親だけimputeされていてもう片親はknown

from VCF import VCFSmall
from GenoRecord import GenoRecord
from VCFFamily import VCFFamilyRecord
from VCFImputable import VCFImputable
from ImputerByParentProgeny import ImputerByParentProgeny
from Map import *


#################### VCFOneParentProgenyImputed ####################

class VCFOneParentProgenyImputed(VCFImputable):
	def __init__(self, samples: list[str], records: list[VCFFamilyRecord],
					ref_haps: list[list[int]], should_impute_mat: bool,
					map_: Map, vcf: VCFSmall):
		VCFImputable.__init__(self, samples, vcf)
		self.records: list[VCFFamilyRecord] = records
		self.imputer = ImputerByParentProgeny(records, ref_haps,
												should_impute_mat, map_, 0.01)
		self.should_impute_mat = should_impute_mat
	
	##### virtual methods for VCFGenoBase #####
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def get_records(self) -> list[GenoRecord]:
		return [ r for r in self.records ]
	
	##### virtual methods for VCFFamilyBase #####
	def get_family_record(self, i: int) -> VCFFamilyRecord:
		return self.records[i]
	
	##### virtual methods for VCFImputable #####
	def impute(self) -> None:
		self.imputer.impute()

File: src/VCFOneParentProgenyImputed.cpp
#include <cmath>
#include "../include/VCFOneParentProgenyImputed.h"
#include "../include/ParentProgenyImputer.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFOneParentProgenyImputed ////////////////////

VCFOneParentProgenyImputed::VCFOneParentProgenyImputed(
									const STRVEC& s,
									const std::vector<VCFFamilyRecord *>& rs,
									const std::vector<std::vector<int>>& ref_hs,
									bool should_impute_mat_, const Map& map_,
									double w, const VCFSmall *vcf) :
					VCFImputable(s, vcf),
					records(rs),
					ref_haps(ref_hs),
					imputer(records, ref_hs, should_impute_mat_, map_, w),
					should_impute_mat(should_impute_mat_) { }

VCFOneParentProgenyImputed::~VCFOneParentProgenyImputed() {
	Common::delete_all(records);
}

size_t VCFOneParentProgenyImputed::amount() const {
	const size_t	NH = ref_haps.size();
	const size_t	P = samples.size() - 2;
	return NH * (NH+P*2+1) << (P*2+1);
}

STRVEC VCFOneParentProgenyImputed::imputed_samples() const {
	STRVEC	ss(samples.size()-2);
	const size_t	index = should_impute_mat ? 0 : 1;
	ss[0] = samples[index];
	std::copy(samples.begin() + 3, samples.end(), ss.begin() + 1);
	return ss;
}

void VCFOneParentProgenyImputed::impute() {
	imputer.impute();
}
